jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerContextGenerator.java

```diff
@@ -36,26 +37,91 @@
  *
- * <p>The image consists of a base image layer and three application layers under the directories:
+ * <p>The image consists of a base image layer and 5 application layers under the directories:
  *
  * <ul>
- *   <li>libs/ (dependency jars)
- *   <li>resources/
- *   <li>classes
+ *   <li>{@code libs/} (dependency jars)
+ *   <li>{@code snapshot-libs/} (snapshot dependency jars)
+ *   <li>{@code resources/} (resource files)
+ *   <li>{@code classes/} ({@code .class} files)
+ *   <li>{@code root/} (extra files)
  * </ul>
+ *
+ * Empty application layers are omitted.
  */
 public class DockerContextGenerator {
 
-  private final SourceFilesConfiguration sourceFilesConfiguration;
+  private static final String DEPENDENCIES_LAYER_DIRECTORY = "libs";
+  private static final String SNAPSHOT_DEPENDENCIES_LAYER_DIRECTORY = "snapshot-libs";
+  private static final String RESOURCES_LAYER_DIRECTORY = "resources";
+  private static final String CLASSES_LAYER_DIRECTORY = "classes";
+  private static final String EXTRA_FILES_LAYER_DIRECTORY = "root";
+
+  /** Represents a Dockerfile {@code COPY} directive. */
+  private static class CopyDirective {
+
+    /** The source files to put into the context. */
+    private final ImmutableList<Path> sourceFiles;
+
+    /** The directory in the context to put the source files for the layer */
+    private final String directoryInContext;
+
+    /** The extraction path in the image. */
+    private final String extractionPath;
+
+    private CopyDirective(
+        ImmutableList<Path> sourceFiles, String directoryInContext, String extractionPath) {
+      this.sourceFiles = sourceFiles;
+      this.directoryInContext = directoryInContext;
+      this.extractionPath = extractionPath;
+    }
+  }
+
+  /**
+   * Adds a copy directive for the {@code layerEntry} if it's not empty.
+   *
+   * @param listBuilder the {@link ImmutableList.Builder} to add to
+   * @param layerEntry the layer entry
+   * @param directoryInContext the directory in the context to put the source files for the layer
+   */
+  private static void addIfNotEmpty(
+      ImmutableList.Builder<CopyDirective> listBuilder,
+      LayerEntry layerEntry,
+      String directoryInContext) {
+    if (layerEntry.getSourceFiles().isEmpty()) {
+      return;
+    }
+
+    listBuilder.add(
+        new CopyDirective(
+            layerEntry.getSourceFiles(), directoryInContext, layerEntry.getExtractionPath()));
+  }
+
+  private final ImmutableList<CopyDirective> copyDirectives;
 
   @Nullable private String baseImage;
   private List<String> jvmFlags = Collections.emptyList();
   private String mainClass = "";
   private List<String> javaArguments = Collections.emptyList();
   private List<String> exposedPorts = Collections.emptyList();
 
-  public DockerContextGenerator(SourceFilesConfiguration sourceFilesConfiguration) {
-    this.sourceFilesConfiguration = sourceFilesConfiguration;
+  // TODO: Just take the LayerConfigurations.
```

Comment:
Would it be that difficult to just do this now? I think we should avoid adding TODOs while trying to fix other issues.

Reply:
I'll look into that again, but the main issue is that since this is in `jib-core`, it doesn't know what layer label corresponds to which copy directive.